src/MonoMod.Core/Platforms/Runtimes/FxCoreBaseRuntime.cs

```diff
@@ -161,6 +162,62 @@ public virtual RuntimeMethodHandle GetMethodHandle(MethodBase method)
+            var declaringType = method.DeclaringType ?? typeof(object);
+            if (!method.IsGenericMethod && !declaringType.IsGenericType)
+                return false;
+
+            // If the method itself is generic, check whether at least one
+            // of its generic arguments makes it a shared instantiation.
+            var methodGenericArguments = method.IsGenericMethod ? method.GetGenericArguments() : Type.EmptyTypes;
+            if (methodGenericArguments.Any(IsGenericSharedType))
+                return true;
+
+            // If the method is effectively generic (i.e., it is defined on a generic type),
+            // we need to determine whether it still requires a hidden argument in the form
+            // of a MethodDesc, MethodTable, or TypeHandle pointer, or if `this` alone will
+            // suffice for this purpose.
+            //
+            // The rules are as follows:
+            //  - If both the method and its declaring type are generic, `this` alone cannot
+            //    provide the generic context the VM needs. In such cases,
+            //    we require a hidden argument for shared instantiations.
+            //  - If the method is static, there is obviously no `this` from which
+            //    the VM could infer the generic context.
+            //  - If `this` is a value type, the unboxed `this` pointer, by definition,
+            //    does not have an associated MethodTable pointer.
+            //  - If the method is a default interface method called via interface dispatch,
+            //    the VM cannot use `this` to derive the generic context because it is too
+            //    ambiguous (e.g., it may implement multiple IFoo<T> interfaces).
+            //
+            // See:
+            // https://github.com/dotnet/runtime/blob/55eee324653e01cf28809d02b25a5b0894b58d22/src/coreclr/vm/method.cpp#L1649
+            var mayNeedHiddenArg = method.IsGenericMethod
+                || method.IsStatic
+                || declaringType.IsValueType
+                || declaringType.IsInterface && !method.IsAbstract;
+            if (!mayNeedHiddenArg)
+                return false;
+
+            // Finally, check whether the type on which the method is defined may be shared.
+            var typeGenericArguments = declaringType.IsGenericType ? declaringType.GetGenericArguments() : Type.EmptyTypes;
+            return typeGenericArguments.Any(IsGenericSharedType);
+        }
+
+        private static bool IsGenericSharedType(Type type)
+        {
+            // Primitives are never shared.
+            // Value types are shared only if at least one of their nested generic arguments is shared.
+            // All other types (i.e., reference types) are always shared.
+            return (
+                !type.IsPrimitive &&
+                (!type.IsValueType || type.IsGenericType && type.GetGenericArguments().Any(IsGenericSharedType))
+            );
+        }
```

Comment:
It might be worth caching this somewhere, possibly in a `CWT<Type, StrongBox<bool>>`? Esp. seeing as it can recurse arbitrarily deeply. Not sure how important that is though.

Reply:
While it *theoretically* can recurse quite deeply, I don't think it's worth optimizing for this scenario in practice. For this method to become substantially computationally expensive, we would need to encounter a deeply nested generic value type in the form of `Struct0<Struct1<Struct2<Struct3<...>, ...>, ...>, ...>` with no reference-type generic arguments to break the loop. And even then, how deep would it really need to be to cause a problem? Even a hundred levels of nesting would be processed in an instant; and in all the years I've worked with .NET, the worst atrocities I've encountered haven't gone beyond "just" a few dozens of total generic arguments.

This is such an insanely rare *(if not outright unrealistic)* scenario that, to be quite frankly honest with you, if someone does create types like this, they deserve to face the consequences of their actions! :D

Given this, I feel like polluting the runtime with a permanent cache for something we might not even encounter a second time would be a significantly worse tradeoff than simply computing the result of this relatively cheap operation as is.

Reply:
I have written such code, though in general I do agree.